### PR TITLE
Implementation of `hardcode_no_ct()`, `hardcode_ct()`, `assign_no_ct()` and `assign_ct()`

### PR DESCRIPTION
**Pull request by ramiromagno**
Thank you for your Pull Request! We have developed this task checklist from the
[Development Process
Guide](https://pharmaverse.github.io/sdtm.oak/CONTRIBUTING.html)
to help with the final steps of the process. Completing the below tasks helps to
ensure our reviewers can maximize their time on your code as well as making sure
the oak codebase remains robust and consistent.

Please check off each taskbox as an acknowledgment that you completed the task
or check off that it is not relevant to your Pull Request. This checklist is
part of the Github Action workflows and the Pull Request will not be merged into
the `devel` branch until you have checked off each task.

- [x] Place Closes #<insert_issue_number> into the beginning of your Pull z
Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the 
[tidyverse style guide](https://style.tidyverse.org/). Run 
`styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should
consider realistic data scenarios and edge cases, e.g. empty datasets, errors,
boundary cases etc. - See 
[Unit Test Guide](https://pharmaverse.github.io/admiraldev/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you 
fully follow the 
[deprecation guidance](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#deprecation)?
- [x] Update to all relevant roxygen headers and examples, including keywords
and families. Refer to the 
[categorization of functions](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the
`NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` if the changes pertain to a user-facing function (i.e. it
has an `@export` tag) or documentation aimed at users (rather than developers)
- [x] Build oak site `pkgdown::build_site()` and check that all affected
examples are displayed correctly and that all new functions occur on the "[Reference](https://pharmaverse.github.io/admiral/reference/index.html)" page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

### CODE REVIEW

File: R/hardcode_no_ct.R [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+hardcode_no_ct <- function(raw_dataset,
+                           raw_variable,
+                           target_sdtm_variable,
+                           target_hardcoded_value,
+                           target_dataset = raw_dataset,
+                           merge_to_topic_by = NULL) {
+  dplyr::right_join(x = raw_dataset, y = target_dataset, by = merge_to_topic_by) |>
+    dplyr::mutate("{raw_variable}" := overwrite(!!rlang::sym(raw_variable), target_hardcoded_value)) |>
+    dplyr::rename("{target_sdtm_variable}" := raw_variable) |>
+    dplyr::relocate(target_sdtm_variable, .after = dplyr::last_col())
```
Comment (rammprasad) [inline R/hardcode_no_ct.R:68]:
```suggestion
  sdtm_ds <- raw_dataset |>
    dplyr::select(rlang::sym(raw_variable), merge_to_topic_by) |>
    dplyr::mutate(target_sdtm_variable = ifelse(!is.na(raw_dataset[[raw_variable]]),
                                                target_hardcoded_value,
                                                NA_character_)) |>
    dplyr::rename("{target_sdtm_variable}" := target_sdtm_variable)
  
  sdtm_ds <- dplyr::left_join(target_dataset, sdtm_ds, by = merge_to_topic_by) |>
    dplyr::select(-rlang::sym(raw_variable))
```

File: R/hardcode_no_ct.R [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+#'   then this parameter defines keys to use in the join between `raw_dataset`
+#'   and `target_dataset`.
+#'
+#' @examples
+#' MD1 <-
+#'   tibble::tribble(
+#'     ~oak_id, ~raw_source, ~patient_number, ~MDRAW,
+#'     1L, "MD1", "PATNUM", "BABY ASPIRIN",
+#'     2L, "MD1", "PATNUM", "CORTISPORIN",
+#'     3L, "MD1", "PATNUM", NA_character_,
```
Comment (rammprasad) [inline R/hardcode_no_ct.R:21]: PATNUM can be numbers like 101, 102, 103 etc

File: R/hardcode_no_ct.R (hunk 1 above)
Comment (rammprasad) [reply R/hardcode_no_ct.R:68]: This logic appears to be ok, but we should add more assertions to ensure its validity. Of course, there is always room for improvement.

**Review by rammprasad — COMMENTED** [review]
Great start, Ramiro.

File: R/hardcode_no_ct.R (hunk 1 above)
Comment (ramiromagno) [reply R/hardcode_no_ct.R:68]: Hi Ramm,

Thank you for the feedback!

May I ask, if the logic is okay, why then the suggested changes? They seem to yield equivalent code to the one I already have, except for one difference: 

- when no argument is passed to `target_dataset`, then it defaults to `raw_dataset`, and with your suggestions that means you will drop all variables from the target dataset except for the join keys (`merge_to_topic_by`) and the variable to be mutated, i.e. `target_sdtm_variable`. Is this intended? I think not.

Moreover, in my code I use this helper function `overwrite()` which nicely wraps that if-else pattern to preserve `NA` (in your suggestion) when hardcoding values. I guess this function may be useful more times as this feels like is going to be recurrent pattern. I've also written a kind of sibling function `rewrite()` which nicely recodes values (in the same spirit of `dplyr::recode()`/`dplyr::case_when()`), and that I expect to be useful when dealing with controlled terminology mappings.

One more question: except for the join keys, should the variables in the raw and target data sets be mutually exclusive? Should I assert for this possibility, and trigger an error if there is at least one variable in common between the two data sets?

Comment (edgar-manukyan) [reply R/hardcode_no_ct.R]: In roak sometimes we notice that the same record gets value from two different topics causing VAR.x and VAR.y creation. In this case we halt the process and message to the user. Hopefully with this nice modular approach we wont see this happening. Perhaps having an assertion won't hurt.

- https://github.com/pharmaverse/roak_pilot/blob/main/R/oak_coalesce_non_joined_duplicate.R

File: R/hardcode.R [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+  # TODO: Assertions.
+
+  tgt_val <- ct_map(tgt_val, ct = ct, cl = cl)
+
+  raw_dat |>
+    dplyr::right_join(y = tgt_dat, by = by) |>
+    dplyr::mutate("{tgt_var}" := recode(x = !!rlang::sym(raw_var), to = tgt_val)) |>
+    dplyr::select(-rlang::sym(raw_var)) |>
+    dplyr::relocate(tgt_var, .after = dplyr::last_col())
```
Comment (edgar-manukyan) [inline R/hardcode.R:20]: Isn't the derived variable already the last one?

**Review by edgar-manukyan — COMMENTED** [review]
Thank you so much @ramiromagno, very impressive work and smart reuse of coding logic. 💯 ⭐ ⭐ ⭐ 

I left few suggestions in the code and more questions for @rammprasad :)

Some items here as well:
- Assertions, just a kind reminder.
- Types of the parameters in the roxygen header, just a kind reminder.
- In the tests can we please have a ct dataset with more than one unqiue value
of codelist_code? 🙏

File: R/assign.R [hunk 4]
```diff
@@ -0,0 +1,388 @@
...
+#' Derive an SDTM variable
+#'
+#' @description
+#' - [assign_no_ct()] maps a variable in a source dataset to a target SDTM
+#' variable that has no terminology restrictions.
+#'
+#' - [assign_ct()] maps a variable in a source dataset to a target SDTM variable
+#' following controlled terminology recoding.
+#'
+#' @param raw_dataset The raw dataset.
```
[the remark this answers is not in the document]
Comment (ramiromagno) [reply R/assign.R:29]: Of course, as soon as I have your thumbs up on the function working as intended, I will clarify parameters, do units tests, assertions, etc..

File: R/ct.R [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+ct_map <-
+  function(x,
+           ct = NULL,
+           cl = NULL,
+           from =  c("collected_value", "term_synonyms"),
+           to = "term_value") {
+
+    ct %||% return(x)
+
+    cl <- cl %||% unique(ct$codelist_code)
```
[the remark this answers is not in the document]
Comment (ramiromagno) [reply R/ct.R:93]: I did it such that the function behavior becomes general, perhaps too general :) The idea is that if the user does not specify a code list code, then all remappings available in `ct` are performed. Not sure if this use case is ever useful... perhaps you prefer that if `cl` is `NULL` then no recoding is performed (as we do when `ct` is `NULL`)...? If you could provide example inputs and outputs for this function, then I can make the function behavior more precise in these corner cases.

File: R/ct.R [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+  ct_mappings <-
+    ct |>
+    dplyr::mutate(to = !!rlang::sym(to)) |>
+    tidyr::pivot_longer(cols = dplyr::all_of(cols),
+                        values_to = "from",
+                        names_to = "type") |>
+    dplyr::select(c("type", "from", "to")) |>
+    dplyr::mutate(type = factor(.data$type, levels = cols)) |>
+    dplyr::arrange(.data$type) |>
```
[the remark this answers is not in the document]
Comment (ramiromagno) [reply R/ct.R:49]: Sure, again, this was not done because I wasn't sure the approach for controlled terminology was okay. Is it?

File: R/ct.R [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    dplyr::select(-"type") |>
+    tidyr::drop_na(.data$from) |>
+    dplyr::mutate(from = str_split(.data$from)) |>
```
[the remark this answers is not in the document]
Comment (ramiromagno) [reply R/ct.R:52]: Sure, will do.

[the remark this answers is not in the document]
Comment (ramiromagno) [reply R/ct.R:52]: Yes, there is. It won't handle quoted strings that might have `";"` inside and that should not be split.

File: R/hardcode.R (hunk 3 above)
Comment (ramiromagno) [reply R/hardcode.R:20]: Not always, namely when the derived variable is already existing and might be placed in some other order other than the last.

File: R/hardcode.R [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+#' @param target_sdtm_variable_codelist_code A codelist code indicating which
+#'   subset of the controlled terminology to apply in the derivation.
+#'
+#' @returns The target dataset with the derived variable `target_sdtm_variable`.
+#'
+#' @examples
+#' MD1 <-
+#'   tibble::tribble(
+#'     ~oak_id, ~raw_source, ~patient_number, ~MDRAW,
+#'     1L, "MD1", 101L, "BABY ASPIRIN",
```
[the remark this answers is not in the document]
Comment (ramiromagno) [reply R/hardcode.R:52]: yes, let me see how that plays out with lintr.

File: R/recode.R [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+#' )
+#'
+#' @keywords internal
+recode <- function(x,
+                    from = unique(na.omit(x)),
+                    to = from,
+                    .no_match = x,
+                    .na = NA) {
+  # to <- rlang::rep_along(x, to)
+  to <- vctrs::vec_recycle(to, length(from))
```
[the remark this answers is not in the document]
Comment (ramiromagno) [reply R/recode.R:87]: Not always. I also use `recode()` in the hardcode family functions, where we are typically recoding one single value. This way it is a bit more general.

File: R/assign.R [hunk 10]
```diff
@@ -0,0 +1,388 @@
...
+#' @param raw_variable The raw variable.
+#' @param target_sdtm_variable The target SDTM variable.
+#' @param target_dataset Target dataset. By default the same as `raw_dataset`.
+#' @param merge_to_topic_by If `target_dataset` is different than `raw_dataset`,
+#'   then this parameter defines keys to use in the join between `raw_dataset`
+#'   and `target_dataset`.
+#' @param study_ct Study controlled terminology specification.
```
[the remark this answers is not in the document]
Comment (ramiromagno) [reply R/assign.R:218]: I think your question hints at a more deep issue here. I think we should have a proper S3 (or even the new S7) classed object that represents controlled terminology mappings. It could be self-validating in the sense of ensuring that there are always the needed variables involved in the recoding. Then, functions like `assign_ct()` would only need to assert that `ct` is indeed an object of a certain type, such that all these other assumptions about having this and that column could be taken for granted.

File: R/assign.R [hunk 11]
```diff
@@ -0,0 +1,388 @@
...
+                        tgt_dat = raw_dat,
+                        by = NULL,
+                        ct = NULL,
+                        cl = NULL) {
+
+  # TODO: Assertions.
+
+  raw_dat |>
+    dplyr::right_join(y = tgt_dat, by = by) |>
+    dplyr::mutate("{tgt_var}" := ct_map(!!rlang::sym(raw_var), ct = ct, cl = cl)) |>
```
[the remark this answers is not in the document]
Comment (ramiromagno) [reply R/assign.R:14]: Could you provide an example in this context? And another one when the `tgt_dat` is not supplied.

File: R/ct.R (hunk 5 above)
[the remark this answers is not in the document]
Comment (ramiromagno) [reply R/ct.R:93]: In the meeting of today we decided that it could indeed be potentially useful to support this use case of not providing a code list code.

File: R/assign.R (hunk 11 above)
[the remark this answers is not in the document]
Comment (ramiromagno) [reply R/assign.R]: As per Ramm's changes this does not apply anymore.

File: R/hardcode.R (hunk 3 above)
Comment (ramiromagno) [reply R/hardcode.R]: Not applicable anymore.

File: R/hardcode_no_ct.R (hunk 1 above)
Comment (ramiromagno) [reply R/hardcode_no_ct.R]: hi @edgar-manukyan, could you check whether this issue still applies with the most recent code?

File: R/hardcode.R [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+                          ct = NULL,
+                          cl = NULL) {
+
+  # TODO: Assertions.
+
+  tgt_val <- ct_map(tgt_val, ct = ct, cl = cl)
+
+  raw_dat |>
+    dplyr::right_join(y = tgt_dat, by = by) |>
+    dplyr::mutate("{tgt_var}" := recode(x = !!rlang::sym(raw_var), to = tgt_val)) |>
```
[the remark this answers is not in the document]
Comment (ramiromagno) [reply R/hardcode.R:67]: you mean checking whether `tgt_var` exists in `tgt_dat`?

File: R/assign.R [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+#' derived variable, as indicated in `tgt_var`.
+#'
+#'
+#' @importFrom rlang :=
+#' @keywords internal
+sdtm_assign <- function(raw_dat,
+                        raw_var,
+                        tgt_var,
+                        ct = NULL,
+                        cl = NULL,
```
Comment (rammprasad) [inline R/assign.R:42]: Does this name sound ok?
cod_dat <- Study codelist specification dataset (ct)
cod_lst --> Code List (cl)

File: R/assign.R [hunk 14]
```diff
@@ -0,0 +1,243 @@
...
+    raw_var = raw_var,
+    tgt_var = tgt_var,
+    tgt_dat = tgt_dat,
+    id_vars = id_vars
+  )
+}
+
+#' @order 2
+#' @export
+#' @rdname assign
```
Comment (rammprasad) [inline R/assign.R:217]: Should we add examples for `assign_ct` as well?

**Review by rammprasad — COMMENTED** [review]
Minor comments.

File: R/hardcode.R [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+#' @name harcode
+NULL
+
+#' @export
+#' @rdname harcode
+hardcode_no_ct <- function(raw_dataset,
+                           raw_variable,
+                           target_sdtm_variable,
+                           target_hardcoded_value,
+                           target_dataset = raw_dataset,
```
[the remark this answers is not in the document]
Comment (ramiromagno) [reply R/hardcode.R]: This comment is no longer applicable as the function interface has changed since.

File: R/assign.R (hunk 13 above)
Comment (ramiromagno) [reply R/assign.R:42]: How far do you want to go with the renaming? I mean there are now several places in the code where `ct` and `cl` are used.

File: R/assign.R (hunk 14 above)
Comment (ramiromagno) [reply R/assign.R:217]: There are already. Take a look at line 177.

File: R/assign.R (hunk 13 above)
Comment (rammprasad) [reply R/assign.R:42]: All the places will be good. We can keep the parameter names consistent.

Comment (rammprasad) [reply R/assign.R:42]: What do you think about this?

cl_spec - Study codelist specification dataset (ct)
cl_name - Codelist name, example - C12345
read_cl_spec_example()
read_cl_spec()
assert_cl_spec()
assert_cl_name()

File: R/assign.R [hunk 16]
```diff
@@ -0,0 +1,243 @@
+#' Derive an SDTM variable
+#'
+#' @description
+#' [sdtm_assign()] is an internal function packing the same functionality as
+#' [assign_no_ct()] and [assign_ct()] together but aimed at developers only.
+#' As a user please use either [assign_no_ct()] or [assign_ct()].
+#'
```
Comment (rammprasad) [inline R/assign.R:7]: We need to add test cases for these functions. We can add it in a new branch.

Code coverage
```R
sdtm.oak Coverage: 67.52%
R/assign.R: 0.00%
R/clear_cache.R: 0.00%
R/hardcode.R: 0.00%
R/oak_id_vars.R: 0.00%
R/zzz.R: 0.00%
R/dtc_problems.R: 55.26%
R/dtc_utils.R: 60.53%
R/ct.R: 79.25%
R/dtc_create_iso8601.R: 89.00%
R/parse_dttm_fmt.R: 91.87%
R/derive_study_day.R: 91.89%
R/dtc_parse_dttm.R: 100.00%
R/recode.R: 100.00%
R/str_split.R: 100.00%
```

**Review by rammprasad — APPROVED** [review]
Works amazing. Minor comments. Ready to merge.

**Comment by rammprasad** [comment]
@edgar-manukyan @ShiyuC - Ready for the second review.

File: R/assign.R (hunk 16 above)
Comment (ramiromagno) [reply R/assign.R:7]: I've created a new issue for the improvement of code coverage.

File: R/recode.R [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+#'   .na = "missing"
+#' )
+#'
+#' @keywords internal
+recode <- function(x,
+                    from = unique(na.omit(x)),
+                    to = from,
+                    .no_match = x,
+                    .na = NA) {
+  # to <- rlang::rep_along(x, to)
```
[the remark this answers is not in the document]
Comment (ramiromagno) [reply R/recode.R]: Outdated now.

File: R/hardcode_no_ct.R (hunk 1 above)
Comment (ramiromagno) [reply R/hardcode_no_ct.R]: I think this is outdated now.

File: R/ct.R (hunk 7 above)
[the remark this answers is not in the document]
Comment (ramiromagno) [reply R/ct.R:240]: Hi @edgar-manukyan: let me know if you think this is still neeed.

File: DESCRIPTION [hunk 18]
```diff
@@ -38,11 +38,15 @@ Depends: R (>= 4.2)
 Imports: 
     admiraldev,
     dplyr (>= 1.0.0),
+    memoise,
```
Comment (edgar-manukyan) [inline DESCRIPTION:41]: At Roche we had a bad experience with memoise, though in our case we had to cached dozens of large datasets. Let's keep this for now and see hot it goes.

**Review by edgar-manukyan — APPROVED** [review]
Thanks so much @ramiromagno for an amazing work. 💯 ⭐ ⭐ 

Looks good to me and left some nice to have comments and thanks for opening issue #48 to increase the coverage 🙏

File: DESCRIPTION (hunk 18 above)
Comment (ramiromagno) [reply DESCRIPTION:41]: Yeah, let us use it judiciously. :)

File: R/ct.R [hunk 19]
```diff
@@ -0,0 +1,433 @@
...
+#'   )
+#'
+#' # Load a controlled terminology example
+#' (ct_spec <- read_ct_spec_example("ct-01-cm"))
+#'
+#' # Use all possible matching terms in the controlled terminology.
+#' ct_map(x = terms, ct_spec = ct_spec)
+#'
+#' # Note that if the controlled terminology mapping is restricted to a codelist
+#' # code, e.g. C71113, then only `"/day"` gets mapped to `"QD"`; remaining terms
```
[the remark this answers is not in the document]
Comment (ramiromagno) [reply R/ct.R]: Done.
